simple-local-avatars.php

```diff
@@ -1,571 +1,610 @@
+	 */
+	public function avatar_defaults( $avatar_defaults ) {
+		remove_action( 'get_avatar', array( $this, 'get_avatar' ) );
+		return $avatar_defaults;
+	}
+
+	/**
+	 * Delete avatars based on a user_id
+	 *
+	 * @param int $user_id
+	 */
+	public function avatar_delete( $user_id ) {
+		$old_avatars = (array) get_user_meta( $user_id, 'simple_local_avatar', true );
+
+		if ( empty( $old_avatars ) )
+			return;
+
+		// if it was uploaded media, don't erase the full size or try to erase an the ID
+		if ( array_key_exists( 'media_id', $old_avatars ) )
+			unset( $old_avatars['media_id'], $old_avatars['full'] );
+
+		if ( ! empty( $old_avatars ) ) {
+			$upload_path = wp_upload_dir();
+
+			foreach ($old_avatars as $old_avatar ) {
+				// derive the path for the file based on the upload directory
+				$old_avatar_path = str_replace( $upload_path['baseurl'], $upload_path['basedir'], $old_avatar );
+				if ( file_exists( $old_avatar_path ) )
+					unlink( $old_avatar_path );
+			}
+		}
+
+		delete_user_meta( $user_id, 'simple_local_avatar' );
+		delete_user_meta( $user_id, 'simple_local_avatar_rating' );
+	}
+
+	/**
+	 * Creates a unique, meaningful file name for uploaded avatars.
+	 *
+	 * @param string $dir Path for file
+	 * @param string $name Filename
+	 * @param string $ext File extension (e.g. ".jpg")
+	 * @return string Final filename
+	 */
+	public function unique_filename_callback( $dir, $name, $ext ) {
+		$user = get_user_by( 'id', (int) $this->user_id_being_edited );
+		$name = $base_name = sanitize_file_name( $user->display_name . '_avatar_' . time() );
+
+		// ensure no conflicts with existing file names
+		$number = 1;
+		while ( file_exists( $dir . "/$name$ext" ) ) {
+			$name = $base_name . '_' . $number;
+			$number++;
+		}
+
+		return $name . $ext;
+	}
+
+	/**
+	 * Adds errors based on avatar upload problems.
+	 *
+	 * @param WP_Error $errors Error messages for user profile screen.
+	 */
+	public function user_profile_update_errors( WP_Error $errors ) {
+		$errors->add( 'avatar_error', $this->avatar_upload_error );
+	}
+
+	/**
+	 * Registers the simple_local_avatar field in the REST API.
+	 */
+	public function register_rest_fields() {
+		register_rest_field( 'user', 'simple_local_avatar', array(
+			'get_callback' => array( $this, 'get_avatar_rest' ),
+			'update_callback' => array( $this, 'set_avatar_rest' ),
+			'schema' => array(
+				'description' => 'The users simple local avatar',
+				'type' => 'object',
+			)
+		));
+	}
+
+	/**
+	 * Returns the simple_local_avatar meta key for the given user
+	 * @param object $profileuser User object
+	 * @param array|string $input Passed input values to sanitize
+	 */
+	public function get_avatar_rest( $user ) {
+		$local_avatar = get_user_meta( $user[ 'id' ], 'simple_local_avatar', true );
+		if ( empty($local_avatar) ) {
+			return;
+		}
+		return $local_avatar;
+	}
+
+	/**
+	 * Updates the simple local avatar from a REST request.
```

Comment:
I think it would be good to note here in the doc block what the built-in REST permissions check is so that in the future we don't wonder why we aren't worried about checking user caps or nonces. Or should we be worried about that?